common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -39,17 +86,735 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+     * Main differences with standard implementation in [LocalMSALController] are:
+     * - No scopes are passed in as part of the (developer provided) parameters. The scopes from the
+     * AT are used to make the refresh token call.
+     * - When the RT is expired or the refresh token call fails, a different exception is thrown.
+     */
+    @Throws(
+        IOException::class,
+        ClientException::class,
+        ArgumentException::class,
+        ServiceException::class
+    )
     fun acquireTokenSilent(
         parameters: AcquireTokenNoFixedScopesCommandParameters
     ): AcquireTokenResult {
         LogSession.logMethodCall(TAG, "${TAG}.acquireTokenSilent")
 
         val acquireTokenSilentResult = AcquireTokenResult()
-        
+
+        // Validate original AcquireTokenNoScopesCommandParameters parameters
+        parameters.validate()
+
+        // Convert AcquireTokenNoScopesCommandParameters into SilentTokenCommandParameters
+        // so we can re-use existing MSAL logic
+        val silentTokenCommandParameters =
+            CommandUtil.convertAcquireTokenNoFixedScopesCommandParameters(
+                parameters
+            )
+        // Not adding any (default) scopes, because we want to retrieve all tokens (regardless of scope).
+        // Scopes will be added later in the flow, if necessary.
+        val targetAccount: AccountRecord = getCachedAccountRecord(silentTokenCommandParameters)
+
+        // Build up params for Strategy construction
+        val authScheme = silentTokenCommandParameters.authenticationScheme
+        val strategyParameters = OAuth2StrategyParameters.builder()
+            .platformComponents(parameters.platformComponents)
+            .authenticationScheme(authScheme)
+            .build()
+        val strategy = silentTokenCommandParameters.authority.createOAuth2Strategy(strategyParameters)
+
+        val tokenCache = silentTokenCommandParameters.oAuth2TokenCache
+        val cacheRecords = tokenCache.loadWithAggregatedAccountData(
+            silentTokenCommandParameters.clientId,
+            parameters.applicationIdentifier,
+            parameters.mamEnrollmentId,
+            null, // TODO see where else this is needed
+            targetAccount,
+            authScheme
+        ) as List<ICacheRecord>
+
+        // The first element is the 'fully-loaded' CacheRecord which may contain the AccountRecord,
+        // AccessTokenRecord, RefreshTokenRecord, and IdTokenRecord... (if all of those artifacts exist)
+        // subsequent CacheRecords represent other profiles (projections) of this principal in
+        // other tenants. Those tokens will be 'sparse', meaning that their AT/RT will not be loaded
+        val fullCacheRecord = cacheRecords[0]
+
+        if (accessTokenIsNull(fullCacheRecord)) {
+            throw ServiceException(ErrorStrings.NATIVE_AUTH_NO_ACCESS_TOKEN_FOUND, "No access token found during refresh - user must be signed out.", null)
```

Comment:
Wondering if we can loop this exception into the existing return model for Native Auth, where we try to avoid throwing exceptions in favor of return error result objects. @SammyO thoughts?

Reply:
Also "user must be signed out" could be confusing, dev might understand this as they must sign out the account to succeed the silent request, which I doubt is what we are trying to communicate here. Perhaps something like "There's no signed in account" would be more fitting.

Reply:
"Wondering if we can loop this exception into the existing return model for Native Auth, where we try to avoid throwing exceptions in favor of return error result objects" - we don't actually return a state or error result object in the `getAccessToken()` method. It only returns an AccountResult, or an exception is thrown.
We're going to change this in a future version of our SDK, once we add support for MFA, CA, etc.

"dev might understand this as they must sign out the account to succeed the silent request" that's correct, that's what we're trying to communicate. There is no resolution for this error, besides signing in a user again, which can only happen after a signout.

Reply:
@shahzaibj  We can't, as we need the scopes from the access token to set in the refresh token call.

Reply:
Are you doing the exact same thing on iOS too? I think it is important to have consistency across platforms for these kinds of things

Reply:
We are. We have a ticket on our backlog to introduce this functionality on both platforms. https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2724081